bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java

```diff
@@ -0,0 +1,495 @@
+            logger.debug("Current unique address is: {}", properties.get(PROPERTY_UNIQUE_ADDRESS));
+            return;
+        }
+
+        ModbusEndpointThingHandler handler = getEndpointThingHandler();
+        if (handler == null) {
+            return;
+        }
+        getThing().setProperty(PROPERTY_UNIQUE_ADDRESS, handler.getUID().getAsString() + ":" + block.address);
+    }
+
+    /**
+     * Dispose the binding correctly
+     */
+    @Override
+    public void dispose() {
+        tearDown();
+    }
+
+    /**
+     * Unregister the poll task and release the endpoint reference
+     */
+    private void tearDown() {
+        unregisterPollTask();
+        unregisterEndpoint();
+    }
+
+    /**
+     * Returns the current slave id from the bridge
+     */
+    public int getSlaveId() {
+        return slaveId;
+    }
+
+    /**
+     * Get the endpoint handler from the bridge this handler is connected to
+     * Checks that we're connected to the right type of bridge
+     *
+     * @return the endpoint handler or null if the bridge does not exist
+     */
+    private @Nullable ModbusEndpointThingHandler getEndpointThingHandler() {
+        Bridge bridge = getBridge();
+        if (bridge == null) {
+            logger.debug("Bridge is null");
+            return null;
+        }
+        if (bridge.getStatus() != ThingStatus.ONLINE) {
+            logger.debug("Bridge is not online");
+            return null;
+        }
+
+        ThingHandler handler = bridge.getHandler();
+        if (handler == null) {
+            logger.debug("Bridge handler is null");
+            return null;
+        }
+
+        if (handler instanceof ModbusEndpointThingHandler) {
+            ModbusEndpointThingHandler slaveEndpoint = (ModbusEndpointThingHandler) handler;
+            return slaveEndpoint;
+        } else {
+            logger.debug("Unexpected bridge handler: {}", handler);
+            return null;
+        }
+    }
+
+    /**
+     * Get a reference to the modbus endpoint
+     */
+    private void connectEndpoint() {
+        if (endpoint.isPresent()) {
+            return;
+        }
+
+        ModbusEndpointThingHandler slaveEndpointThingHandler = getEndpointThingHandler();
+        if (slaveEndpointThingHandler == null) {
+            @SuppressWarnings("null")
+            String label = Optional.ofNullable(getBridge()).map(b -> b.getLabel()).orElse("<null>");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE,
+                    String.format("Bridge '%s' is offline", label));
+            logger.debug("No bridge handler available -- aborting init for {}", label);
+            return;
+        }
+
+        try {
+            slaveId = slaveEndpointThingHandler.getSlaveId();
+
+            @Nullable
+            ModbusSlaveEndpoint optionalEndpoint = slaveEndpointThingHandler.asSlaveEndpoint();
```

Comment:
This whole logic seems to be very complicated, compared to a classic null-check.

```
endpoint = slaveEndpointThingHandler.asSlaveEndpoint();
if (endpoint != null) {
   ...
}
```

looks much clearer. If you want to stay with the Optional:

```
  endpoint = Optional.ofNullable(slaveEndpointThingHandler.asSlaveEndpoint());
```

Reply:
Unfortunately your second alternative does not satisfy the eclipse null checker: the Optional.ofNullable has the following signature:
```
Optional<T> Optional.ofNullable(T value)
```
and thus it assumes that T in this cases is also annotated with @NonNull

So far this was the only place where the Optional made me problem compared to the @Nullable approach